Implementation Plan: Filtering and designating sensitive results in the API

sarayourfriend opened the pull request:
## Fixes

<!-- If PR doesn't fully resolve the issue, replace 'Fixes' below with 'Related to'. -->
<!-- If there is no issue being resolved, please consider opening one before creating this pull request. -->

Fixes #937 by @sarayourfriend 

Opened 30 March, **due 13 April**.

## Description

<!-- Concisely describe what the pull request does. -->
<!-- Add screenshots, videos, or other media to show the problem and the solution when appropriate. -->
I've requested reviews from Staci and Dhruv as the IP touches Airflow, API, and ingestion server changes. Staci and Dhruv are both knowledgeable in at least one or several of these areas. If either of y'all are unable to review this, please let me know and I'll ask someone else.

## Testing Instructions

<!-- Give steps for the reviewer to verify that this PR fixes the problem; or delete this section entirely. -->
Read the IP. I do not have any outstanding questions, but please focus on the plan and make sure that it makes sense and is feasible. If you have suggestions for how to break any steps up into separate PRs, please share those especially.

## Checklist

<!-- Replace  the [ ] with [x] to check the boxes. -->

- [x] My pull request has a descriptive title (not a vague title like
      `Update index.md`).
- [x] My pull request targets the _default_ branch of the repository (`main`) or
      a parent feature branch.
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [N/A] I added or updated tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [N/A] I tried running the project locally and verified that there are no visible
      errors.

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<!-- You must read and understand the following attestation. -->

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

zackkrida commented on rfcs/detecting_sensitive_textual_content/20230308-implementation_plan_filtering_and_designating_results_with_sensitive_textual_content.md:
```diff
@@ -0,0 +1,458 @@
+
+## Project links
+
+[Project Thread](https://github.com/WordPress/openverse/issues/377)
+[Project Proposal](./20230309-project_proposal_detecting_sensitive_textual_content.md)
+[Milestone (pending)](#)
+[Implementation Plan: Managing A Sensitive Terms List](https://github.com/WordPress/openverse/pull/911)
```
```suggestion
- [Project Thread](https://github.com/WordPress/openverse/issues/377)
- [Project Proposal](./20230309-project_proposal_detecting_sensitive_textual_content.md)
- [Milestone (pending)](#)
- [Implementation Plan: Managing A Sensitive Terms List](https://github.com/WordPress/openverse/pull/911)
```

nit for readibility

stacimc commented on rfcs/detecting_sensitive_textual_content/20230308-implementation_plan_filtering_and_designating_results_with_sensitive_textual_content.md:
```diff
@@ -0,0 +1,458 @@
+documents marked mature _and_ those with sensitive textual content. Critically
+this means that results in the origin index that are mature but do not have
+sensitive text would get marked as having sensitive text based on their presence
+in the `sensitve_result_ids` set.
```
Just double checking my understanding:  the goal is to make it possible to determine if an API result was marked sensitive due to `provider_supplied_mature` vs matching `sensitive_text`, but (at least for now) they will be treated exactly the same.

stacimc replied:
Ah, that makes a lot of sense! Thank you for that link.

AetherUnbound commented on rfcs/detecting_sensitive_textual_content/20230308-implementation_plan_filtering_and_designating_results_with_sensitive_textual_content.md:
```diff
@@ -0,0 +1,458 @@
+   that do not have sensitive terms in the three textual content fields
+   (`title`, `description`, `tags.name`). This is the filtered index.
+   - Requires the following:
+     - SENSITIVE_TERMS injected as new environment variable. See
```
Nit:
```suggestion
     - `SENSITIVE_TERMS` injected as new environment variable. See
```

stacimc commented on rfcs/detecting_sensitive_textual_content/20230308-implementation_plan_filtering_and_designating_results_with_sensitive_textual_content.md:
```diff
@@ -0,0 +1,458 @@
+
+Because the new functionality will all be either internal (ingestion server and
+catalogue) or behind an unstable query parameter, we do not need any new feature
+flags to manage the rollout of this work.
```
Since `include_sensitive_results` is by default False, even though the actual query parameter is unstable won't this be immediately rolled out in production once the first data refresh including the new index changes completes?

Do we have a planned way to roll back/query from the original index in case of some problem with the filtered index in production? Is that necessary?

sarayourfriend replied:
> Do we have a planned way to roll back/query from the original index in case of some problem with the filtered index in production? Is that necessary?

That's a good point, I hadn't considered the mismatch there. Let's add a Django setting for enabling the new query parameter.

AetherUnbound commented on rfcs/detecting_sensitive_textual_content/20230308-implementation_plan_filtering_and_designating_results_with_sensitive_textual_content.md:
```diff
@@ -0,0 +1,458 @@
+   entries based on the results of the previous step and the existing `mature`
+   field.
+
+Many thanks to @robfelty for helping me find a much simpler approach to
+sensitive result designation than I had originally devised. This solution is
+easily 10x more intelligible and simpler to document and implement than the
+others I had considered.
```
Agreed, this makes a lot more sense on initial read to my Elasticsearch-novice experience!

AetherUnbound commented on rfcs/detecting_sensitive_textual_content/20230308-implementation_plan_filtering_and_designating_results_with_sensitive_textual_content.md:
```diff
@@ -0,0 +1,458 @@
+
+1. Create the new filtered index before calling reindex, using the
+   [`index_settings`](https://github.com/wordpress/openverse/blob/38af21c359703d3faee9e160f1ac69372661d6a2/ingestion_server/ingestion_server/es_mapping.py#L1)
+   function to configure this index in the same was as the origin index for the
```
```suggestion
   function to configure this index in the same way as the origin index for the
```

AetherUnbound commented on rfcs/detecting_sensitive_textual_content/20230308-implementation_plan_filtering_and_designating_results_with_sensitive_textual_content.md:
```diff
@@ -0,0 +1,458 @@
+   - In this option we can rely on the reindex API to create the new index as ES
+     will apply the template settings automatically.
+
+The first is my recommendation as it matches our current approach and is trivial
+to implement. The code sample at the end of this section demonstrates it in the
+call to `es.indices.create`.
+
+The second option, using index templates, is an interesting alternative that we
+could explore in the future, especially once we remove the ingestion server.
+However, it breaks significantly from our current approach to index
+configuration and I do not think is worth pursuing as part of this project.
```
Agreed!

AetherUnbound commented on rfcs/detecting_sensitive_textual_content/20230308-implementation_plan_filtering_and_designating_results_with_sensitive_textual_content.md:
```diff
@@ -0,0 +1,458 @@
+def create_and_populate_filtered_index(
+  self, model_name: str, index_suffix: str, **_
+):
+    source_index = f"{model_name}-{index_suffix}"
+    destination_index = f"{source_index}-filtered"
+
+    self.es.indices.create(
+        index=destination_index,
+        body=index_settings(model_name),
+    )
+
+    self.es.reindex(
+        body={
+            "source": {
+                "index": source_index,
+                "query": {
+                    "bool": {
+                        "must_not": [
+                            {
+                                "multi_match": {
+                                    "query": f'"{term}"',
+                                    "fields": ["tags.name", "title", "description"],
+                                }
+                            }
+                            for term in SENSITIVE_TERMS
+                        ]
+                    }
+                },
+            },
+            "dest": {"index": destination_index},
+        },
+        slices="auto",
+        wait_for_completion=True,
+    )
+
+    self.refresh(index_name=destination_index, change_settings=True)
+
+    self.ping_callback()
```
So straightforward!! 😮

AetherUnbound commented on rfcs/detecting_sensitive_textual_content/20230308-implementation_plan_filtering_and_designating_results_with_sensitive_textual_content.md:
```diff
@@ -0,0 +1,458 @@
+   to be searched. ES only automatically refreshes indexes that are being
+   actively used, and the index should be refreshed before it starts being used.
+
+The creation of the new index follow the same pattern of creating the regular
```
```suggestion
The creation of the new index follows the same pattern of creating the regular
```

AetherUnbound commented on rfcs/detecting_sensitive_textual_content/20230308-implementation_plan_filtering_and_designating_results_with_sensitive_textual_content.md:
```diff
@@ -0,0 +1,458 @@
+
+> **Note**
+>
+> While both processes have a "copy" step, they are notably different. The
+> origin index receives its documents from Postgres via ingestion workers, which
+> push the documents into Elasticsearch. The filtered index receives its
+> documents via the reindex API and does not require the ingestion workers.
```
Thanks for pointing out this distinction 🙂 We should make sure that the code path you describe above doesn't spin up the indexer worker instances since they won't be needed.

sarayourfriend replied:
What would be required to ensure that? Assuming we don't call the `schedule_distributed_index` function, it should be fine, right? I don't see anywhere else that would cause the workers to get spun up.

AetherUnbound replied:
I think that's it! Awesome

AetherUnbound commented on rfcs/detecting_sensitive_textual_content/20230308-implementation_plan_filtering_and_designating_results_with_sensitive_textual_content.md:
```diff
@@ -0,0 +1,458 @@
+
+### Query the filtered index (overview step 3)
+
+We must add a new boolean query parameter to the search query serialiser,
+`unstable__include_sensitive_results`. This parameter will default to `False`.
+We will remove the `unstable` designation during the wrap-up of this
+implementation plan. This parameter should also reflect the state of the
+`mature` parameter: when `mature` is `True`, the new parameter should also be
+`True`. This prepares us to deprecate the `mature` parameter when we remove the
+`unstable` designation from the new parameter.
+
+When `unstable__include_sensitive_results` is `False` (default behaviour), query
+the filtered index by appending the `-filtered` suffix to the index name. Update
+the existing `mature` parameter handling to check
+`unstable__include_sensitive_results`. When the new parameter is `False`,
+results with the `mature` set to `true` should also be filtered out.
+
+When `unstable__include_sensitive_results` is `True`, query the origin index
+(existing behaviour).
+
+### Derive the list of sensitive document IDs based on the filtered index (overview step 4)
+
+When `unstable__include_sensitive_results` is `True`, we need to derive which
+documents in the results have sensitive textual content. To do this, we will
+rely on the fact that those documents are _not_ in the filtered index.
+Therefore, we can pull the list of `_id`s for the documents retrieved from the
+origin index and query for their presence in the filtered index. If they are in
+the filtered index, then we know they do not have sensitive textual content.
+Something along the following lines of pseudocode:
```
This is very much a nit, but there's some oscillation between `True`/`true` and `False`/`false` here - is there any importance behind that difference? If not, would it make sense to make them consistent across the document?

sarayourfriend replied:
No reason other than missing the shift bar a couple times :slightly_smiling_face: I'll consolidate to using the Python title cased version.

sarayourfriend replied:
Ah, actually, the difference between `true` and `True` here _was_ intentional! In Elasticsearch the "mature" property is a JSON field, so the value is indeed `true` with a lowercase T. I'm still changing the language to not reference the specific values, but I just wanted to note that there is indeed a small difference here due to the data representation model that Elasticsearch is actually using.

stacimc commented on rfcs/detecting_sensitive_textual_content/20230308-implementation_plan_filtering_and_designating_results_with_sensitive_textual_content.md:
````diff
@@ -0,0 +1,458 @@
+    "point_alias": {"alias": f"{target_alias}-filtered"}
+}
+```
+
````
Checking I understand:  any changes to the `SENSITIVE_TERMS` list (addition or removal of terms) will not be reflected in search results until the next data refresh, right?

I'm hoping the data refresh will become much more stable soon so this won't be a problem🤞, but if we needed to it sounds like we could also manually trigger just the relevant `create_and_populate_filtered_index` and `point_alias` steps on the ingestion server to refresh these results without a full data refresh.

Do you think it would be worth having the steps to create the filtered index in a separate DAG? We could then trigger the DAG from the data refresh using a [TriggerDagRunOperator](https://docs.astronomer.io/learn/cross-dag-dependencies#triggerdagrunoperator) so that it would continue to be an automatic part of standard data refreshes, but we could also run it independently. We _would_ need to make sure the new DAG can't be manually triggered while a data refresh is already running, which would require a bit of tweaking.

stacimc replied:
Exactly, I remembered talking about a one-week timeframe being acceptable at some point, but the instability of the data refresh might change that 😞 

> To prevent them running at the same time, is it a matter of adding some sort of lock in Airflow variables? I vaguely recall you'd done some work on another DAG to prevent this as well.

Yes, the data refresh DAGs themselves use a `data_refresh` pool with one slot and a custom `SingleRunExternalDAGSensor` to make sure that only one data refresh DAG runs at a time. We could add these new DAGs to the list of "data refresh DAGs", but that wouldn't be quite right by default -- because the 'parent' DAG will still be running when the filtered index creation DAG starts.

I think the `TriggerDagRunOperator` will allow us to [pass in conf options](https://registry.astronomer.io/providers/apache-airflow/versions/latest/modules/triggerdagrunoperator) when we trigger the filtered index creation DAG, so we should be able to pass a "force" parameter like you suggested. Then we would probably want a new, simple sensor that would `wait` as long as the data refresh is running, but also accept the "force" parameter as an override.

There might be a simpler way to do it, which I can look into, but I think that ought to work!

sarayourfriend replied:
> Then we would probably want a new, simple sensor that would wait as long as the data refresh is running

The data refresh takes so long: would it make more sense for the dag to fail if called during the data refresh rather than waiting? If the only time that kind of error can happen is if we manually call it (this dag itself won't be scheduled, right?) then waiting would be more confusing to me rather than an error that says "I can't do this while the data refresh is going".

stacimc replied:
For sure, we could definitely fail the DAG instead. Now you mention it, the data refresh will include the filtered index creation step as is... so there shouldn't theoretically be any reason to wait and re-do it after the data refresh finishes, anyway. That sounds great 👍

AetherUnbound commented on rfcs/detecting_sensitive_textual_content/20230308-implementation_plan_filtering_and_designating_results_with_sensitive_textual_content.md:
````diff
@@ -0,0 +1,458 @@
+slippery concept without concrete examples (at least that's what I found when
+writing this document).
+
+```
````
```suggestion
```py
```

AetherUnbound commented on rfcs/detecting_sensitive_textual_content/20230308-implementation_plan_filtering_and_designating_results_with_sensitive_textual_content.md:
````diff
@@ -0,0 +1,458 @@
+origin result IDs in the final set subtraction, we'd end up with the following
+situation:
+
+```
````
```suggestion
```py
```

AetherUnbound commented on rfcs/detecting_sensitive_textual_content/20230308-implementation_plan_filtering_and_designating_results_with_sensitive_textual_content.md:
````diff
@@ -0,0 +1,458 @@
+same: we will exclude results that are both marked mature and have sensitive
+textual content:
+
+```
````
```suggestion
```py
```

AetherUnbound commented on rfcs/detecting_sensitive_textual_content/20230308-implementation_plan_filtering_and_designating_results_with_sensitive_textual_content.md:
```diff
@@ -0,0 +1,458 @@
+  {id: 1, mature: false}, # no sensitive text
+]
+
+origin_index_result_ids - filtered_index_result_ids
```
Suggested change for clarity:
```suggestion
sensitive_result_ids = origin_index_result_ids - filtered_index_result_ids
```

AetherUnbound commented on rfcs/detecting_sensitive_textual_content/20230308-implementation_plan_filtering_and_designating_results_with_sensitive_textual_content.md:
```diff
@@ -0,0 +1,458 @@
+origin_index_less_mature_result_ids = [r for r in origin_index results if not r["mature"]]
+# => [1, 3]
+
+origin_index_less_mature_result_ids - filtered_index_result_ids
```
Suggested change for clarity:
```suggestion
sensitive_result_ids = origin_index_less_mature_result_ids - filtered_index_result_ids
```

stacimc commented on rfcs/detecting_sensitive_textual_content/20230308-implementation_plan_filtering_and_designating_results_with_sensitive_textual_content.md:
```diff
@@ -0,0 +1,458 @@
+final list of sensitive results without needing to rely on the diversely
+filtered index. However, this adds a big re-indexing burden (yet another
+additional index) that can be bypassed altogether if we just exclude mature
+results from the filtered index to begin with.
```
Thank you for these examples! The problem is quite clear and your reasoning here makes sense to me.

AetherUnbound commented on rfcs/detecting_sensitive_textual_content/20230308-implementation_plan_filtering_and_designating_results_with_sensitive_textual_content.md:
```diff
@@ -0,0 +1,458 @@
+continue to operate as an undocumented alias for the `include_sensitive_results`
+parameter.
+
+## Feature flags
+
+Because the new functionality will all be either internal (ingestion server and
+catalogue) or behind an unstable query parameter, we do not need any new feature
+flags to manage the rollout of this work.
```
I might be misunderstanding this, doesn't this contradict the `# Feature Flags` section above?

sarayourfriend replied:
Whoops! I'm going to fix this in light of https://github.com/WordPress/openverse/pull/996#discussion_r1162007337 as well.

AetherUnbound commented on rfcs/detecting_sensitive_textual_content/20230308-implementation_plan_filtering_and_designating_results_with_sensitive_textual_content.md:
```diff
@@ -0,0 +1,458 @@
+result 4 was not in the set of IDs from the filtered index _only_ because it is
+marked mature or if it also has sensitive textual content.
+
+The only way I can think of to solve this is to have an index solely of
+documents that _do_ have sensitive textual content and querying it to derive the
+final list of sensitive results without needing to rely on the diversely
+filtered index. However, this adds a big re-indexing burden (yet another
+additional index) that can be bypassed altogether if we just exclude mature
+results from the filtered index to begin with.
```
Forgive me if I'm missing a piece here - could we exclude mature results from the filtered index as well when we're creating it? Would there be a need to include mature results that aren't sensitive separately, especially since we're going to be excluding those results anyway if the sensitive/mature results aren't requested to be included in the final results?

sarayourfriend replied:
> Forgive me if I'm missing a piece here - could we exclude mature results from the filtered index as well when we're creating it? Would there be a need to include mature results that aren't sensitive separately, especially since we're going to be excluding those results anyway if the sensitive/mature results aren't requested to be included in the final results?

As Staci said, the section deals entirely with the idea of excluding mature documents from the filtered index and describes why it's an impossible solution. @AetherUnbound, your inclination is precisely why I wrote the section this paragraph is under, because I also considered it a possibility and anticipated that others would have the same question. If the explanation is not clear on why it's not possible for us to exclude the mature results from the filtered index, please let me know so that I can revise the section.

AetherUnbound replied:
Admittedly I read this section several times and still came away a bit confused 😅 I think the premise was getting to me - my thought was that it doesn't matter for the API whether the `mature` field is `True` or whether we filtered it for sensitive text reasons, the results are excluded all the same. So _in my mind_ making the distinction between `mature` and `sensitive` was unnecessary, but it seems that it might be in some cases which would require the explanation you put out Sara. I guess the core of my question was whether that distinction was important, although you've [answered my question in a later comment](https://github.com/WordPress/openverse/pull/996#discussion_r1162130216)!

I perhaps was getting confused by the distinction between the `mature` API parameter (which we are saying is functionally equivalent to `sensitive` after the implementation of this plan) and the `mature` _field_ on the API (which will stick around in perpetuity) 🤯 I believe I'm all cleared up now!

sarayourfriend replied:
> I perhaps was getting confused by the distinction between the mature API parameter (which we are saying is functionally equivalent to sensitive after the implementation of this plan) and the mature field on the API (which will stick around in perpetuity) exploding_head I believe I'm all cleared up now!

This will be confusing for a while, unfortunately :disappointed:. Hopefully decoupling the parameter name from the various things that it includes will help (i.e., using the broader "sensitive" term to capture "mature" results as well as other types of sensitivity, textual, and machine vision derived).

> So in my mind making the distinction between mature and sensitive was unnecessary, but it seems that it might be in some cases which would require the explanation you put out Sara

I can add another note about this in the section referencing the frontend. The existence of the `sensitivity` field as an array that notes the distinction hopefully communicates the necessity of knowing the difference between them. The reason for that need isn't explicit in this document because the project proposal covers the reason we are making this distinction.

AetherUnbound reviewed (approved):
This is an excellent IP, thanks for the work you've done in exploring the options here! I know I'm not a reviewer but I was curious about how this would turn out on the Elasticsearch side so I gave it a look. Feel free to use my approval for the document or wait until the requested reviewers have a chance to review.

All of my commentary is either surface/syntax changes or questions, nothing blocking!

stacimc reviewed (commented):
This is fantastic work and a very interesting read :) Thanks very much for the code samples and examples, this was quite easy to follow even as a newcomer to Elasticsearch!

sarayourfriend commented on rfcs/detecting_sensitive_textual_content/20230308-implementation_plan_filtering_and_designating_results_with_sensitive_textual_content.md:
```diff
@@ -0,0 +1,458 @@
+       filtered index
+     - Updates to the `load_sample_data.sh` script to generate the filtered
+       index
+     - A benign list of terms to use for local testing that matches results in
```
Yes, good point. Do you think we should maintain the testing list external from the application, rather than as the default value when reading the environment variable?

stacimc replied:
+1, I think that would be the safest approach. Great call out!

sarayourfriend replied:
To clarify, you think that we should _not_ just use the default value for the environment variable?

If we don't then we'd need to update `env.docker` for the ingestion server and copy those to the ingestion server settings.

If we set the default in code then it is (in my opinion) slightly easier to understand in testing. We can restrict the default to only apply not in production:

```py
SENSITIVE_TERMS = config(
    "SENSITIVE_TERMS",
    cast=lambda x: x.split("\n"),
    default=( # just an example list
        """
        water
        dogs
        cactus
        """
        if ENVIRONMENT != "production" else None
    )
)
```

We can raise an error in `create_and_populate_filtered_index` if `SENSITIVE_TERMS` is None.

stacimc replied:
The addition of the environment check works just fine and I agree, easier to understand. Your example here is totally fine with me.

sarayourfriend commented on rfcs/detecting_sensitive_textual_content/20230308-implementation_plan_filtering_and_designating_results_with_sensitive_textual_content.md:
```diff
@@ -0,0 +1,458 @@
+    "ingest_upstream": {},
+    "promote": {"alias": target_alias},
+    "create_and_populate_filtered_index": {},
+    "point_alias": {"alias": f"{target_alias}-filtered"}
```
Yes! We can't use `promote` for that reason exactly, because it also promotes the API table, which isn't necessary for the filtered index.

stacimc replied:
💯 Perfect! This is great, I love how easily this fits into the existing flow.

sarayourfriend commented on rfcs/detecting_sensitive_textual_content/20230308-implementation_plan_filtering_and_designating_results_with_sensitive_textual_content.md:
```diff
@@ -0,0 +1,458 @@
+implementation plan. This parameter should also reflect the state of the
+`mature` parameter: when `mature` is `True`, the new parameter should also be
+`True`. This prepares us to deprecate the `mature` parameter when we remove the
+`unstable` designation from the new parameter.
```
I hadn't considered this. Personally I'd rather 400 requests with conflicting parameters.

stacimc replied:
I think you're right, even better not to assume either way.

dhruvkb commented on rfcs/detecting_sensitive_textual_content/20230308-implementation_plan_filtering_and_designating_results_with_sensitive_textual_content.md:
```diff
@@ -0,0 +1,458 @@
+   that do not have sensitive terms in the three textual content fields
+   (`title`, `description`, `tags.name`). This is the filtered index.
+   - Requires the following:
+     - SENSITIVE_TERMS injected as new environment variable. See
```
This is a very small detail but I'm wondering if a file stored in S3 and it's URL in the environment variable would be a better option that loading the entire list of sensitive terms from an environment variable. It's not an immediate issue but if we include sensitive terms from multiple languages, the list might be several megabytes which might exceed some character and size limits.

sarayourfriend replied:
I had thought about doing it that way as well, and I don't remember why I landed on using an environment variable. Alternatively, we could even request the list directly from GitHub, as it will be a public repository storing them.

The difficulty this presents is for testing data that doesn't need to be stored in a public repository. We could either have an `inane-terms.txt` in the repository for testing or use `SENSITIVE_TERMS_URL` as the primary and `SENSITIVE_TERMS` as a fallback that would only be used in non-production environments.

dhruvkb reviewed (approved):
LGTM! I have one concern about the use of env vars to store the terms list but everything else about the implementation seems well-thought out and nicely planned to me.

sarayourfriend commented:
@AetherUnbound Thank you for the editorial suggestions. I've applied them in https://github.com/WordPress/openverse/pull/996/commits/830f9a747b595362880bcdf117ae8965268a991f. Rather than noting that in each suggestion, I've gone ahead and just resolved the conversations, but I wanted to make sure you knew I was grateful for your suggestions and did apply them. I've also added the clarifications we discussed.

@stacimc I've made the changes we talked about to the feature flags, conflicting parameter handling, and the Airflow DAG configurations. @AetherUnbound I would appreciate you careful review of the DAG situation as well :pray: 

@dhruvkb I made the change you suggested to use a URL instead of putting the list in the environment variables. It's a much more flexible solution that just generally makes more sense. I also added instructions for how to handle this in staging and local environments by using a locally served file from Falcon itself. This seemed like the easiest approach without adding too much in the code for handling fallbacks of different types of settings for different environments.

AetherUnbound commented on rfcs/detecting_sensitive_textual_content/20230308-implementation_plan_filtering_and_designating_results_with_sensitive_textual_content.md:
```diff
@@ -0,0 +1,581 @@
+DAGs, we will create a separate DAG that will be triggered by the data refresh
+DAGs using the
+[`TriggeredDagRunOperator`](https://airflow.apache.org/docs/apache-airflow/stable/_api/airflow/operators/trigger_dagrun/index.html).
+The new DAG will be named `<media_type>_filtered_index_creation`[^passive-name].
+It will not run on a schedule (only triggered), it will have a max concurrency
+of one, and it will accept the following configuration variables:
+
+[^passive-name]:
+    The passive voice name of this DAG is not ideal, but
+    `create_filtered_<model_name>_index_dag` reads as a DAG factory function
+    rather than a dag named "create filtered index".
```
We don't need `_dag` in the suffix, do you think `create_filtered_<model_name>_index` works? That sounds fine IMO

sarayourfriend replied:
Great! I will remove this then :sweat_smile:

AetherUnbound commented on rfcs/detecting_sensitive_textual_content/20230308-implementation_plan_filtering_and_designating_results_with_sensitive_textual_content.md:
```diff
@@ -0,0 +1,581 @@
+### Query the filtered index (overview step 3)
+
+We must add a new boolean query parameter to the search query serialiser,
+`unstable__include_sensitive_results`. This parameter will default the disabled
+state. We will remove the `unstable` designation during the wrap-up of this
```
```suggestion
`unstable__include_sensitive_results`. This parameter will default to the disabled
state. We will remove the `unstable` designation during the wrap-up of this
```

AetherUnbound commented on rfcs/detecting_sensitive_textual_content/20230308-implementation_plan_filtering_and_designating_results_with_sensitive_textual_content.md:
```diff
@@ -0,0 +1,581 @@
+before it's ready to delete the previous index, we don't want to play with race
+conditions.
+
+Rather than trying to use a lock, we will read the DAG status directly from the
+DAG and fail the filtered index creation DAG run if the corresponding data
```
Can you clarify which DAG `the DAG` is here? I think it's the data refresh DAG but clarity would be helpful:
```suggestion
Rather than trying to use a lock, we will read the DAG status directly from the
data refresh DAG and fail the filtered index creation DAG run if the corresponding data
```

sarayourfriend replied:
I think the wording is confusing. I'll change it to "we will read the data refresh DAG status at the start of the create filtered index DAG" to remove the ambiguity.

AetherUnbound commented on rfcs/detecting_sensitive_textual_content/20230308-implementation_plan_filtering_and_designating_results_with_sensitive_textual_content.md:
```diff
@@ -0,0 +1,581 @@
+   and wait for completion
+   - See existing data refresh DAG strategy for making and waiting for ingestion
+     server requests
+1. Call the `point_alias`[^no-promote-action] action in the ingestion server and
+   wait for completion.
+1. Delete the previous filtered index whose canonical name we retrieved in the
+   first step (`DELETE_INDEX` action)
+
+[^no-promote-action]:
+    We cannot use the `promote` action used by the regular data refresh flow
+    because while that action does point the alias to the newly created index,
+    it also promotes the API tables, which is not necessary for filtered index
+    creation. Therefore, we use the `point_alias` action which only points
+    aliases and nothing else.
```
Forgive my pedantry here, but should the various actions all be capitalized? The mixed case is throwing me off in thinking that they might be referring to different things:
```suggestion
1. Call the `POINT_ALIAS`[^no-promote-action] action in the ingestion server and
   wait for completion.
1. Delete the previous filtered index whose canonical name we retrieved in the
   first step (`DELETE_INDEX` action)

[^no-promote-action]:
    We cannot use the `PROMOTE` action used by the regular data refresh flow
    because while that action does point the alias to the newly created index,
    it also promotes the API tables, which is not necessary for filtered index
    creation. Therefore, we use the `POINT_ALIAS` action which only points
    aliases and nothing else.
```

sarayourfriend replied:
For clarity yes. These actions correspond 1-to-1 with the `TableIndexer` methods, but the API does ask for them in caps.

AetherUnbound reviewed (approved):
Thanks for making the changes suggested 🙂 I have a few more points of clarification specifically around the changes. No blockers, this looks good to me 👍🏼

sarayourfriend commented:
Thanks @AetherUnbound. Address your comments in https://github.com/WordPress/openverse/pull/996/commits/a1fad8c575d91ba6fc58c7e6608c78804cbcf137

stacimc reviewed (approved):
Fantastic! +1 for injecting the `SENSITIVE_TERMS_URL` as an environment variable. The updates around the DAG all look great to me and document the concurrency problems very clearly.

This looks great, @sarayourfriend!

sarayourfriend commented on rfcs/detecting_sensitive_textual_content/20230308-implementation_plan_filtering_and_designating_results_with_sensitive_textual_content.md:
```diff
@@ -0,0 +1,574 @@
+
+## Reviewers
+
+- [ ] TBD
+- [ ] TBD
```
```suggestion
- [x] @stacimc
- [x] @dhruvkb
```
